.github/workflows/build.yml

```diff
@@ -99,29 +99,92 @@ jobs:
-        if: steps.test_env.outputs.type == 'plugin'
+      - if: ${{ steps.cache-images.outputs.cache-hit == 'true' }}
+        name: Load saved docker images
         run: |
-          # download keycloak cas provider
-          sudo wget https://github.com/jacekkow/keycloak-protocol-cas/releases/download/18.0.2/keycloak-protocol-cas-18.0.2.jar -O /opt/keycloak-protocol-cas-18.0.2.jar
-
-          ./ci/pod/openfunction/build-function-image.sh
-          make ci-env-up project_compose_ci=ci/pod/docker-compose.${{ steps.test_env.outputs.type }}.yml
-          sudo ./ci/init-${{ steps.test_env.outputs.type }}-test-service.sh
-
-      - name: Start CI env (LAST_TEST)
-        if: steps.test_env.outputs.type == 'last'
+          if [[ -f docker-images-backup/apisix-images.tar ]]; then
+            echo "load saved images..., time: $(date)"
+            docker load --input docker-images-backup/apisix-images.tar
+            echo "load saved images done, time: $(date)"
+            if [[ ${{ steps.test_env.outputs.type }} == 'last' ]]; then
+              # generating SSL certificates for Kafka
+              sudo keytool -genkeypair -keyalg RSA -dname "CN=127.0.0.1" -alias 127.0.0.1 -keystore ./ci/pod/kafka/kafka-server/selfsigned.jks -validity 365 -keysize 2048 -storepass changeit
+            fi
+            if [[ ${{ steps.test_env.outputs.type }} == 'plugin' ]]; then
+              # download keycloak cas provider
+              sudo wget https://github.com/jacekkow/keycloak-protocol-cas/releases/download/18.0.2/keycloak-protocol-cas-18.0.2.jar -O /opt/keycloak-protocol-cas-18.0.2.jar
+            fi
+            make ci-env-up project_compose_ci=ci/pod/docker-compose.${{ steps.test_env.outputs.type }}.yml
+            if [[ ${{ steps.test_env.outputs.type }} != 'first' ]]; then
+              sudo ./ci/init-${{ steps.test_env.outputs.type }}-test-service.sh
+            fi
+            echo "lanuch saved images done, time: $(date)"
+          fi
+      - if: ${{ steps.cache-images.outputs.cache-hit != 'true' }}
+        name: Linux lanuch services
         run: |
-          # generating SSL certificates for Kafka
-          sudo keytool -genkeypair -keyalg RSA -dname "CN=127.0.0.1" -alias 127.0.0.1 -keystore ./ci/pod/kafka/kafka-server/selfsigned.jks -validity 365 -keysize 2048 -storepass changeit
-          make ci-env-up project_compose_ci=ci/pod/docker-compose.${{ steps.test_env.outputs.type }}.yml
-          sudo ./ci/init-${{ steps.test_env.outputs.type }}-test-service.sh
+          if [[ ${{ steps.test_env.outputs.type }} == 'first' ]]; then
+            # launch deps env
+            echo "start first service ..., time: $(date)"
+            make ci-env-up project_compose_ci=ci/pod/docker-compose.${{ steps.test_env.outputs.type }}.yml
+            echo "start first service done, time: $(date)"
+          fi
+          if [[ ${{ steps.test_env.outputs.type }} == 'plugin' ]]; then
+            echo "start plugin service ..., time: $(date)"
```

Comment:
Wouldn't it be better to encapsulate these scripts into a shell file? for example
```shell
# init-plugin-test-service.sh
before() {
           sudo wget https://github.com/jacekkow/keycloak-protocol-cas/releases/download/18.0.2/keycloak-protocol-cas-18.0.2.jar -O /opt/keycloak-protocol-cas-18.0.2.jar
            bash ci/pod/openfunction/build-function-image.sh
}

after() {
            sudo ./ci/init-${{ steps.test_env.outputs.type }}-test-service.sh
            echo "start plugin service done, time: $(date)"
}
```

'first' and 'last' are similar.

Reply:
Try to adjust it like this, the current shell judgment logic is not clear enough. And there are some repeated commands, such as: `sudo keytool -genkeypair -keyalg RSA -dname "CN=127.0.0.1" -alias 127.0.0.1 -keystore ./ci/pod/kafka/kafka-server/selfsigned.jks -validity 365 -keysize 2048 -storepass changeit`

run: |
    sudo ./ci/init-${{ steps.test_env.outputs.type }}-test-service.sh before
    make ci-env-up project_compose_ci=ci/pod/docker-compose.${{ steps.test_env.outputs.type }}.yml
    sudo ./ci/init-${{ steps.test_env.outputs.type }}-test-service.sh after

    bash ./ci/backup-docker-images.sh ${{ steps.test_env.outputs.type }}